pennylane/transforms/decorators.py

```diff
@@ -0,0 +1,442 @@
+
+        @functools.wraps(tape_transform)
+        def make_qfunc_transform(fn):
+            @functools.wraps(fn)
+            def internal_wrapper(*args, **kwargs):
+                tape = make_tape(fn)(*args, **kwargs)
+                tape = tape_transform(tape)
+                return tape.measurements
+
+            return internal_wrapper
+
+    make_qfunc_transform.tape_fn = tape_transform
+    return make_qfunc_transform
+
+
+def qnode_transform(qnode_transform_fn):
+    """Register a new QNode transform.
+
+    Args:
+        qnode_transform_fn (QNode transform): the QNode transform function
+            to register.
+
+            Allowed QNode transforms must be functions of the following form:
+
+            .. code-block:: python
+
+                def qnode_transform(qnode, *args, **kwargs):
+                    ...
+                    return tapes, processing_fn
+
+            That is, the first argument must be the input QNode to transform,
+            and the function must return a tuple ``(list, function)`` containing:
+
+            * A list of new tapes to execute, and
+            * A processing function with signature ``processing_fn(List[float])``
+              which is applied to the flat list of results from the executed tapes.
+
+            If ``tapes`` is empty, then it is assumed no quantum evaluations
+            are required, and ``processing_fn`` will be passed an empty list.
+
+    Returns:
+        function: A hybrid quantum-classical function. Takes the same input arguments as
+        the input QNode.
+
+    **Example**
+
+    Given a simple tape transform, that replaces the :class:`~.CRX` gate with a
+    :class:`~.RY` and :class:`~.CZ` operation,
+
+    .. code-block:: python
+
+        @qml.single_tape_transform
+        def tape_transform(tape, x):
+            for op in tape.operations + tape.measurements:
+                if op.name == "CRX":
+                    qml.RY(op.parameters[0] * qml.math.sqrt(x), wires=op.wires[1])
+                    qml.CZ(wires=op.wires)
+                else:
+                    op.queue()
+
+    lets build a QNode transform that applies this transform twice with different
+    transform parameters to create two tapes, and then sum the results:
+
+    .. code-block:: python
+
+        @qml.qnode_transform
+        def my_transform(qnode, x, y):
+            tape1 = tape_transform(qnode.qtape, x)
+            tape2 = tape_transform(qnode.qtape, y)
+
+            def processing_fn(results):
+                return qml.math.sum(results)
+
+            return [tape1, tape2], processing_fn
+
+    It can then be used to transform an existing QNode:
+
+    .. code-block:: python
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @my_transform(0.7, 0.8)
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.Hadamard(wires=0)
+            qml.CRX(x, wires=[0, 1])
+            return qml.expval(qml.PauliZ(1))
+
+    >>> circuit(0.6)

```

Comment:
Where does the processing function go when this happens?

Reply:
I'm not sure I follow the question?

Reply:
Sorry I didn't phrase that well. The QNode transform `my_transform` should return a list of tapes and a processing function when applied to `circuit`. But here it's just used as a decorator, so where do those return values go? Furthermore, when we execute `circuit`, is the returned value the `expval(PauliZ(1))`, or is it the sum of results from the processing function?

Reply:
Ohhh.

When used as a decorator, it transforms the QNode to one that returns
```python
processing_fn([t.execute for t in tapes])
```

Reply:
Oh that's not what I expected; so when it's *not* used as a decorator, it just returns the tapes and functions, but when it is, it returns the evaluation results?

Reply:
wait, that doesn't sound right, which means I think I answered the wrong question. Maybe we should hop on a call/slack?